File: parad/parad/state.py
"""Sync state tracker — stores version/hash info outside the encrypted DB.

State lives in PARADOX_HOME/<name>.sync.json so it survives pull operations
(which replace the entire DB file).
"""

import json
import re
from datetime import datetime, timezone
from pathlib import Path

import parad.config as _config

# Characters that are unsafe in a filename segment.  Multi-part keys like
# "myproject/mydb" are flattened to "myproject__mydb".
_UNSAFE_STATE_CHARS = re.compile(r'[/\\:*?"<>|\x00-\x1f]')


def sanitize_state_key(db_key: str) -> str:
    """Convert a possibly multi-part db key into a safe filename segment.

    ``"myproject/mydb"`` -> ``"myproject__mydb"``; already-safe keys like
    ``"myproject__mydb"`` or ``"main"`` pass through unchanged.
    """
    key = _UNSAFE_STATE_CHARS.sub("__", str(db_key)).strip().strip(".")
    if not key or not any(c.isalnum() for c in key):
        raise ValueError(f"db_key must not be empty after sanitizing: {db_key!r}")
    return key


def _state_path(db_name: str) -> Path:
    # Read CONFIG_DIR from the config module at call time so a runtime
    # PARADOX_HOME override (test suites, embedded use) is always honored.
    return _config.config_dir() / f"{sanitize_state_key(db_name)}.sync.json"


def load_state(db_name: str) -> dict:
    path = _state_path(db_name)
    if path.exists():
        try:
            return json.loads(path.read_text())
        except (json.JSONDecodeError, OSError):
            pass
    return {
        "database_name": db_name,
        "remote_version": None,
        "remote_hash": None,
        "last_sync": None,
        "last_local_hash": None,
        "dirty": False,
        "offline": False,
    }


def save_state(db_name: str, state: dict):
    _config.config_dir().mkdir(parents=True, exist_ok=True)
    _state_path(db_name).write_text(json.dumps(state, indent=2))


def get_remote_version(db_name: str) -> int | None:
    s = load_state(db_name)
    v = s.get("remote_version")
    return int(v) if v is not None else None


def set_remote_version(db_name: str, version: int, file_hash: str = ""):
    s = load_state(db_name)
    s["remote_version"] = version
    s["remote_hash"] = file_hash
    s["last_sync"] = datetime.now(timezone.utc).isoformat()
    save_state(db_name, s)


def get_last_local_hash(db_name: str) -> str | None:
    return load_state(db_name).get("last_local_hash")


def set_last_local_hash(db_name: str, file_hash: str):
    s = load_state(db_name)
    s["last_local_hash"] = file_hash
    save_state(db_name, s)


# ── dirty flag (un-pushed local changes) ────────────────────────────


def mark_dirty(db_key: str):
    s = load_state(db_key)
    s["dirty"] = True
    save_state(db_key, s)


def clear_dirty(db_key: str):
    s = load_state(db_key)
    s["dirty"] = False
    save_state(db_key, s)


def is_dirty(db_key: str) -> bool:
    return bool(load_state(db_key).get("dirty", False))


# ── offline flag (offline → batch push on reconnect) ───────────────


def set_offline(db_key: str, offline: bool):
    s = load_state(db_key)
    s["offline"] = bool(offline)
    save_state(db_key, s)


def is_offline(db_key: str) -> bool:
    return bool(load_state(db_key).get("offline", False))


# ── one-stop status read ────────────────────────────────────────────


def get_sync_status(db_key: str) -> dict:
    """Return the full sync status for a db key as a single dict.

    Keys: ``database_name``, ``remote_version``, ``remote_hash``,
    ``last_sync``, ``last_local_hash``, ``dirty``, ``offline``.
    """
    s = load_state(db_key)
    return {
        "database_name": s.get("database_name", db_key),
        "remote_version": s.get("remote_version"),
        "remote_hash": s.get("remote_hash"),
        "last_sync": s.get("last_sync"),
        "last_local_hash": s.get("last_local_hash"),
        "dirty": bool(s.get("dirty", False)),
        "offline": bool(s.get("offline", False)),
    }

File: client/src/state.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import { configDir } from './config.js';

// Characters unsafe in a filename segment. Multi-part keys like
// "myproject/mydb" are flattened to "myproject__mydb".
// eslint-disable-next-line no-control-regex -- U+0000–U+001F are intentionally blocked as filename-unsafe
const UNSAFE_STATE_CHARS = /[/\\:*?"<>|\u0000-\u001f]/g;

export function sanitizeStateKey(dbKey: string): string {
  const key = dbKey.replace(UNSAFE_STATE_CHARS, '__').trim().replace(/^\.+|\.+$/g, '');
  if (!key || !/[a-zA-Z0-9]/.test(key)) {
    throw new Error(`db_key must not be empty after sanitizing: ${JSON.stringify(dbKey)}`);
  }
  return key;
}

export interface SyncState {
  database_name: string;
  remote_version: number | null;
  remote_hash: string | null;
  last_sync: string | null;
  last_local_hash: string | null;
  dirty: boolean;
  offline: boolean;
}

function defaultState(dbName: string): SyncState {
  return {
    database_name: dbName,
    remote_version: null,
    remote_hash: null,
    last_sync: null,
    last_local_hash: null,
    dirty: false,
    offline: false,
  };
}

function statePath(dbName: string): string {
  return path.join(configDir(), `${sanitizeStateKey(dbName)}.sync.json`);
}

export function loadState(dbName: string): SyncState {
  const p = statePath(dbName);
  try {
    if (fs.existsSync(p)) {
      const parsed = JSON.parse(fs.readFileSync(p, 'utf-8')) as Partial<SyncState>;
      return { ...defaultState(dbName), ...parsed };
    }
  } catch {
    // corrupt/partial state file -> defaults
  }
  return defaultState(dbName);
}

export function saveState(dbName: string, state: SyncState): void {
  const p = statePath(dbName);
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, JSON.stringify(state, null, 2), 'utf-8');
}

export function getRemoteVersion(dbName: string): number | null {
  const s = loadState(dbName);
  const v = s.remote_version;
  return v === null || v === undefined ? null : Number(v);
}

export function setRemoteVersion(dbName: string, version: number, fileHash = ''): void {
  const s = loadState(dbName);
  s.remote_version = version;
  s.remote_hash = fileHash;
  s.last_sync = new Date().toISOString();
  saveState(dbName, s);
}

export function getLastLocalHash(dbName: string): string | null {
  return loadState(dbName).last_local_hash;
}

export function setLastLocalHash(dbName: string, fileHash: string): void {
  const s = loadState(dbName);
  s.last_local_hash = fileHash;
  saveState(dbName, s);
}

export function markDirty(dbKey: string): void {
  const s = loadState(dbKey);
  s.dirty = true;
  saveState(dbKey, s);
}

export function clearDirty(dbKey: string): void {
  const s = loadState(dbKey);
  s.dirty = false;
  saveState(dbKey, s);
}

export function isDirty(dbKey: string): boolean {
  return Boolean(loadState(dbKey).dirty);
}

export function setOffline(dbKey: string, offline: boolean): void {
  const s = loadState(dbKey);
  s.offline = Boolean(offline);
  saveState(dbKey, s);
}

export function isOffline(dbKey: string): boolean {
  return Boolean(loadState(dbKey).offline);
}

export function getSyncStatus(dbKey: string): SyncState {
  return loadState(dbKey);
}
